lib/geminabox/compact_indexer.rb

```diff
@@ -0,0 +1,195 @@
+        File.join(info_path, name)
+      end
+    end
+
+    class Paths
+      include PathMethods
+
+      attr_reader :datadir
+
+      def initialize(datadir)
+        @datadir = datadir
+      end
+    end
+
+    attr_reader :datadir
+
+    def initialize(datadir)
+      @datadir = datadir
+      FileUtils.mkdir_p(info_path)
+    end
+
+    include PathMethods
+
+    def clear_index
+      FileUtils.rm_rf(index_path)
+      FileUtils.mkdir_p(info_path)
+    end
+
+    def fetch_info(name)
+      path = info_name_path(name)
+      File.binread(path) if File.exist?(path)
+    end
+
+    def fetch_versions
+      path = versions_path
+      File.binread(path) if File.exist?(path)
+    end
+
+    def reindex(specs = nil)
+      if specs && File.exist?(versions_path)
+        Gem.time("Compact index incremental reindex") do
+          incremental_reindex(specs)
+        end
+        return
+      end
+
+      Gem.time("Compact index full rebuild") do
+        full_reindex
+      end
+    end
+
+    def yank(spec)
+      with_tmp_dir do |dest_paths|
+        name = spec.name
+        info = DependencyInfo.new(name)
+        data = fetch_info(name)
+        return [] unless data
+
+        info.content = data
+        info.remove_gem_spec(spec)
+
+        prepare_files_for_yank(name, info, dest_paths)
+      end
+    end
+
+    private
+
+    def incremental_reindex(gem_specs)
+      with_tmp_dir do |dest_paths|
+        version_info = VersionInfo.new
+        version_info.load_versions(versions_path)
+
+        updated_files = prepare_info_files(gem_specs, dest_paths, version_info)
+
+        file_path = dest_paths.versions_path
+        updated_files << [file_path, versions_path]
+        version_info.write(file_path)
+
+        updated_files
+      end
+    end
+
+    def prepare_info_files(gem_specs, dest_paths, version_info)
+      updated_files = []
+      gem_specs.group_by(&:name).each do |name, specs|
+        info = DependencyInfo.new(name)
+        data = fetch_info(name)
+        info.content = data if data
+
+        specs.each do |spec|
+          gem_version = GemVersion.from_spec(spec)
+          checksum = Specs.checksum_for_version(gem_version)
+          info.add_gem_spec_and_gem_checksum(spec, checksum)
+        end
+
+        file_path = dest_paths.info_name_path(name)
+        updated_files << [file_path, info_name_path(name)]
+        File.binwrite(file_path, info.content)
+        version_info.update_gem_versions(info)
+      end
+      updated_files
+    end
+
+    def all_specs
+      Geminabox::GemVersionCollection.new(Specs.all_gems)
+    end
+
+    def full_reindex
+      clear_index
+      version_info = VersionInfo.new
+
+      specs = all_specs.by_name.to_h
+      count = specs.size + 1
+      progress_reporter = ui.progress_reporter(count, "Building #{count} compact index files", "Complete")
+      mutex = Mutex.new
+
+      infos = Parallel.map(specs, in_threads: 10) do |name, versions|
```

Comment:
FYI threads in CRuby still do use GIL, so it won't speed up CPU-bound calculations, it could speed up IO-bound operations like writing to a filesystem.
using non-blocking operations like IO.read_nonblock or async should be a better way how to improve IO performance, but I don't think there's a gem with a simple interface like the Parallel gem. At least the maximum number of threads is limited, so I agree this is probably the simplest way to improve it.

Reply:
Yes the GIL is a problem I had almost forgotten about. In fact, the speedup here isn't all that impressive (6 instead of 7 minutes on my laptop for our gem server). But I converted a lot of time back and forth between proxy and standalone to test my conversion script and it helped a bit.